### PR TITLE
Panwang/filewatcher

### PR DESCRIPTION
**Pull request by pan-wang**
remove filewatcher from shim

### CODE REVIEW

File: src/AspNetCoreModuleV2/RequestHandlerLib/application.h [hunk 1]
```diff
@@ -17,9 +18,18 @@ class APPLICATION : public IAPPLICATION
         return m_status;
     }
 
-    APPLICATION()
+    APPLICATION(REQUESTHANDLER_CONFIG* pConfig)
         : m_cRefs(1)
     {
+        m_pConfig = pConfig;
```
Comment (pakrym) [inline src/AspNetCoreModuleV2/RequestHandlerLib/application.h]: Please don't do manual memory management. You can also just pass the application path in.

File: src/AspNetCoreModuleV2/OutOfProcessRequestHandler/dllmain.cxx [hunk 2]
```diff
@@ -21,6 +21,8 @@ IHttpServer *       g_pHttpServer = NULL;
 HINSTANCE           g_hWinHttpModule;
 HINSTANCE           g_hAspNetCoreModule;
 HANDLE              g_hEventLog = NULL;
+FILE_WATCHER*       g_pFileWatcher = NULL;
```
Comment (pakrym) [inline src/AspNetCoreModuleV2/OutOfProcessRequestHandler/dllmain.cxx]: Why do you need it globally? Can't it be per `application` object?

File: src/AspNetCoreModuleV2/InProcessRequestHandler/dllmain.cxx [hunk 3]
```diff
@@ -99,12 +99,12 @@ CreateApplication(
...
         BOOL disableStartupPage = pConfig->QueryDisableStartUpErrorPage();
 
-        auto pApplication = std::make_unique<IN_PROCESS_APPLICATION>(pServer, std::move(config));
+        auto pApplication = std::make_unique<IN_PROCESS_APPLICATION>(pServer, config.release());
         
         if (FAILED(pApplication->LoadManagedApplication()))
         {
             // Set the currently running application to a fake application that returns startup exceptions.
-            *ppApplication = new StartupExceptionApplication(pServer, disableStartupPage);
+            *ppApplication = new StartupExceptionApplication(pServer, config.release(), disableStartupPage);
```
Comment (pakrym) [inline src/AspNetCoreModuleV2/InProcessRequestHandler/dllmain.cxx]: You can't release twice, second time it might be null.

File: src/AspNetCoreModuleV2/AspNetCore/src/applicationinfo.cpp [hunk 4]
```diff
@@ -54,42 +37,119 @@ APPLICATION_INFO::~APPLICATION_INFO()
...
+BOOL APPLICATION_INFO::LoadAppOffline(LPWSTR strFilePath)
+{
+    BOOL            fResult = TRUE;
+    LARGE_INTEGER   li = { 0 };
+    CHAR           *pszBuff = NULL;
+    HANDLE         handle = INVALID_HANDLE_VALUE;
+
+    DBG_ASSERT(strFilePath);
+
+    handle = CreateFile(strFilePath,
```
Comment (pakrym) [inline src/AspNetCoreModuleV2/AspNetCore/src/applicationinfo.cpp]: Handle wrapped and logging in this method.

File: src/AspNetCoreModuleV2/OutOfProcessRequestHandler/outofprocess/outprocessapplication.cpp [hunk 5]
```diff
@@ -47,6 +47,16 @@ OUT_OF_PROCESS_APPLICATION::Initialize(
         }
     }
 
+    if (m_pFileWatcherEntry == NULL)
+    {
+        m_pFileWatcherEntry = new  FILE_WATCHER_ENTRY(g_pFileWatcher);
```
Comment (jkotalik) [inline src/AspNetCoreModuleV2/OutOfProcessRequestHandler/outofprocess/outprocessapplication.cpp]: Shouldn't the file watcher entry be created in the application?

File: src/AspNetCoreModuleV2/AspNetCore/src/applicationinfo.cpp (hunk 4 above)
Comment (pakrym) [reply src/AspNetCoreModuleV2/AspNetCore/src/applicationinfo.cpp]: I did it in https://github.com/aspnet/IISIntegration/pull/977/files#diff-0ecd7731bba35e311fac903916b30e44R26

File: src/AspNetCoreModuleV2/InProcessRequestHandler/dllmain.cxx [hunk 6]
```diff
@@ -99,12 +99,12 @@ CreateApplication(
 
         BOOL disableStartupPage = pConfig->QueryDisableStartUpErrorPage();
 
-        auto pApplication = std::make_unique<IN_PROCESS_APPLICATION>(pServer, std::move(config));
+        auto pApplication = std::make_unique<IN_PROCESS_APPLICATION>(pServer, config.release());
```
Comment (jkotalik) [inline src/AspNetCoreModuleV2/InProcessRequestHandler/dllmain.cxx]: Why was this changed?

**Comment by jkotalik** [comment]
App offline regressed: AppOfflineDroppedWhileSiteRunning_SiteShutsDown. Make sure the site shuts down when app offline is dropped?

File: src/AspNetCoreModuleV2/OutOfProcessRequestHandler/dllmain.cxx (hunk 2 above)
Comment (pan-wang) [reply src/AspNetCoreModuleV2/OutOfProcessRequestHandler/dllmain.cxx]: You don't want to create so many file watcher threads. A global one, single watcher thread

File: src/AspNetCoreModuleV2/OutOfProcessRequestHandler/outofprocess/outprocessapplication.cpp (hunk 5 above)
Comment (pan-wang) [reply src/AspNetCoreModuleV2/OutOfProcessRequestHandler/outofprocess/outprocessapplication.cpp]: I just want to make the application.h simple and abstract enough. an application may not need do file watching for appoffline.

Comment (pakrym) [reply src/AspNetCoreModuleV2/OutOfProcessRequestHandler/outofprocess/outprocessapplication.cpp]: Add `FILE_WATCHING_APPLICATION` then

File: src/AspNetCoreModuleV2/RequestHandlerLib/requesthandler_config.h [hunk 7]
```diff
@@ -209,6 +207,25 @@ class REQUESTHANDLER_CONFIG
         return &m_struConfigPath;
     }
 
+    VOID
+    ReferenceConfig()
```
Comment (jkotalik) [inline src/AspNetCoreModuleV2/RequestHandlerLib/requesthandler_config.h]: Why did you add manual Referencing/Dereferencing to the configuration?

Comment (pan-wang) [reply src/AspNetCoreModuleV2/RequestHandlerLib/requesthandler_config.h]: in the case of inprocess application LoadManagedApplication failed, we then create a failederrorapplication which shares config object with the previous application. As the config object will be owned by the application, we cannot call delete or use smart point, otherwise double free.

Comment (jkotalik) [reply src/AspNetCoreModuleV2/RequestHandlerLib/requesthandler_config.h]: failederrorapplication shouldn't share config with the previous application. Also if that's the case, can't you use a shared pointer?

File: src/AspNetCoreModuleV2/AspNetCore/src/applicationinfo.cpp [hunk 8]
```diff
@@ -54,42 +38,103 @@ APPLICATION_INFO::~APPLICATION_INFO()
...
+APPLICATION_INFO::AppOfflineFound()
 {
-    if (m_pFileWatcherEntry != NULL)
+    ULONGLONG  ulCurrentTime = GetTickCount64();
+    //
+    // we only care about app offline presented. If not, it means the application has started
+    // and is monitoring  the app offline file
+    // we cache the file exist check for 1 minute
+    //
+    if (m_fAppOfflineFound && ulCurrentTime - m_ulLastCheckTime > 1000)
```
Comment (jkotalik) [inline src/AspNetCoreModuleV2/AspNetCore/src/applicationinfo.cpp]: This would not check app_offline on the first request if ulCurrentTime - m_ulLastCheckTime < 1000. For example, if have app_offline already in the folder and then send a request, it will not check for app_offline even though it should.

File: src/AspNetCoreModuleV2/AspNetCore/src/applicationinfo.cpp [hunk 9]
```diff
@@ -54,42 +38,103 @@ APPLICATION_INFO::~APPLICATION_INFO()
...
-    UpdateAppOfflineFileHandle();
 
 Finished:
     return hr;
 }
 
-HRESULT
-APPLICATION_INFO::StartMonitoringAppOffline()
+BOOL
+APPLICATION_INFO::AppOfflineFound()
```
Comment (jkotalik) [inline src/AspNetCoreModuleV2/AspNetCore/src/applicationinfo.cpp]: Rename to CheckIfAppOfflinePresent

File: src/AspNetCoreModuleV2/AspNetCore/Inc/applicationinfo.h [hunk 10]
```diff
@@ -145,16 +132,18 @@ class APPLICATION_INFO
...
     HRESULT FindNativeAssemblyFromHostfxr(HOSTFXR_OPTIONS* hostfxrOptions, PCWSTR libraryName, STRU* location);
+    BOOL    LoadAppOffline(LPWSTR strFilePath);
 
     static VOID DoRecycleApplication(LPVOID lpParam);
 
     mutable LONG            m_cRefs;
     STRU                    m_struInfoKey;
+    STRA                    m_strAppOfflineContent;
     BOOL                    m_fAppOfflineFound;
+    BOOL                    m_fAppOfflineLoaded;
```
Comment (jkotalik) [inline src/AspNetCoreModuleV2/AspNetCore/Inc/applicationinfo.h]: Is m_fAppOfflineLoaded set to false after app_offline is removed?

File: src/AspNetCoreModuleV2/AspNetCore/src/applicationinfo.cpp [hunk 11]
```diff
@@ -98,73 +143,45 @@ APPLICATION_INFO::UpdateAppOfflineFileHandle()
...
-                m_pConfiguration->QueryApplicationPath()->QueryStr())))
-            {
-                UTILITY::LogEvent(g_hEventLog,
-                    EVENTLOG_INFORMATION_TYPE,
-                    ASPNETCORE_EVENT_RECYCLE_APPOFFLINE,
-                    strEventMsg.QueryStr());
-            }
-
-            RecycleApplication();
+            LoadAppOffline(strFilePath.QueryStr());
```
Comment (jkotalik) [inline src/AspNetCoreModuleV2/AspNetCore/src/applicationinfo.cpp]: You never check the return value of LoadAppOffline(). Doesn't this determine m_fAppOfflineFound ?

File: src/AspNetCoreModuleV2/AspNetCore/src/proxymodule.cxx [hunk 12]
```diff
@@ -147,15 +124,21 @@ ASPNET_CORE_PROXY_MODULE::OnExecuteRequestHandler(
...
-        if (pApplication != NULL &&
-            pApplication->QueryStatus() != APPLICATION_STATUS::RUNNING &&
-            pApplication->QueryStatus() != APPLICATION_STATUS::STARTING)
+        DBG_ASSERT(pHttpContext);
+        // make sure application is in running state
+        // cannot recreate the application as we cannot reload clr for inprocess
+        if (pApplication->QueryStatus() == APPLICATION_STATUS::OFFLINE)
+        {
+            pApplicationManager->RecycleApplicationFromManager(m_pApplicationInfo->QueryConfig()->QueryConfigPath()->QueryStr());
+            // todo: add appoffline
```
Comment (jkotalik) [inline src/AspNetCoreModuleV2/AspNetCore/src/proxymodule.cxx]: Is this Todo resolved? Doesn't look like it.

File: src/AspNetCoreModuleV2/RequestHandlerLib/filewatcher.h [hunk 13]
```diff
@@ -1,14 +1,24 @@
...
-#include <windows.h>
+#pragma once
 
-#include "stringu.h"
+#include "stdafx.h"
+#include "iapplication.h"
 
 #define FILE_WATCHER_SHUTDOWN_KEY           (ULONG_PTR)(-1)
 #define FILE_WATCHER_ENTRY_BUFFER_SIZE      4096
+#ifndef CONTAINING_RECORD
```
[the remark this answers is not in the document]
Comment (pakrym) [reply src/AspNetCoreModuleV2/RequestHandlerLib/filewatcher.h]: What about this?

File: src/AspNetCoreModuleV2/AspNetCore/src/proxymodule.cxx (hunk 12 above)
Comment (pan-wang) [reply src/AspNetCoreModuleV2/AspNetCore/src/proxymodule.cxx]: yes. updated it

File: src/AspNetCoreModuleV2/AspNetCore/src/applicationinfo.cpp (hunk 11 above)
Comment (pan-wang) [reply src/AspNetCoreModuleV2/AspNetCore/src/applicationinfo.cpp]: we may fail to load the app_offline content due to file ACL or the file is exclusively opened by some other proces (e.g., antivirus) but the file does exist. this is why we set m_fAppOfflineFound to true. We will return 503 with empty content

File: src/AspNetCoreModuleV2/AspNetCore/Inc/applicationinfo.h (hunk 10 above)
Comment (pan-wang) [reply src/AspNetCoreModuleV2/AspNetCore/Inc/applicationinfo.h]: once appoffline is detected, the instance of applicationinfo will be recycled which may take time. no need to reset m_fAppOfflineLoaded as we don't want bad thing happen in the middle

File: src/AspNetCoreModuleV2/AspNetCore/src/applicationinfo.cpp (hunk 8 above)
Comment (pan-wang) [reply src/AspNetCoreModuleV2/AspNetCore/src/applicationinfo.cpp]: m_ulLastCheckTime is 0, ulCurrentTime is current tick number since machine reboot which cannot be less than 1000 (i.e., 1s)

File: src/AspNetCoreModuleV2/AspNetCore/src/applicationinfo.cpp [hunk 14]
```diff
@@ -54,42 +38,103 @@ APPLICATION_INFO::~APPLICATION_INFO()
...
+
+    if (handle == INVALID_HANDLE_VALUE)
+    {
+        if (HRESULT_FROM_WIN32(GetLastError()) == ERROR_FILE_NOT_FOUND)
+        {
+            fResult = FALSE;
+        }
+
+        // If file is currenlty locked exclusively by other processes, we might get INVALID_HANDLE_VALUE even though the file exists. In that case, we should return TRUE here.
+        goto Finished;
```
[the remark this answers is not in the document]
Comment (pan-wang) [reply src/AspNetCoreModuleV2/AspNetCore/src/applicationinfo.cpp]: OK. This function is the only one do "go finish". In general, I prefer single exit

File: src/AspNetCoreModuleV2/AspNetCore/Inc/applicationmanager.h [hunk 15]
```diff
@@ -97,55 +93,30 @@ class APPLICATION_MANAGER
...
-            if(m_pFileWatcher == NULL)
+            try
             {
-                hr = HRESULT_FROM_WIN32( ERROR_NOT_ENOUGH_MEMORY );
-                goto Finished;
+                m_pApplicationInfoHash = new APPLICATION_INFO_HASH();
             }
-
-            m_pFileWatcher->Create();
+            CATCH_RETURN();
```
[the remark this answers is not in the document]
Comment (pan-wang) [reply src/AspNetCoreModuleV2/AspNetCore/Inc/applicationmanager.h]: No. This initialize happened at RegisterModule which much earlier than ExecuteRequest

File: src/AspNetCoreModuleV2/RequestHandlerLib/filewatcher.h (hunk 13 above)
[the remark this answers is not in the document]
Comment (pan-wang) [reply src/AspNetCoreModuleV2/RequestHandlerLib/filewatcher.h]: A macro used for IIS hashtable

File: src/AspNetCoreModuleV2/AspNetCore/src/applicationinfo.cpp (hunk 8 above)
Comment (pakrym) [reply src/AspNetCoreModuleV2/AspNetCore/src/applicationinfo.cpp]: > we cache the file exist check for 1 minute

One second?

File: src/AspNetCoreModuleV2/AspNetCore/src/applicationinfo.cpp (hunk 11 above)
Comment (pakrym) [reply src/AspNetCoreModuleV2/AspNetCore/src/applicationinfo.cpp]: Why does it have return value at all then?

File: src/AspNetCoreModuleV2/AspNetCore/src/applicationinfo.cpp (hunk 8 above)
Comment (pan-wang) [reply src/AspNetCoreModuleV2/AspNetCore/src/applicationinfo.cpp]: yes. one second. a typo

File: src/AspNetCoreModuleV2/RequestHandlerLib/filewatcher.h (hunk 13 above)
[the remark this answers is not in the document]
Comment (pan-wang) [reply src/AspNetCoreModuleV2/RequestHandlerLib/filewatcher.h]: OK. Remove it.
